marshal/consumer.go

```diff
@@ -235,12 +235,47 @@ func (c *Consumer) defaultTopicPartitions() int {
 	return 0
 }
 
+// claimTerminated is called by a claim when they've terminated. This is used so we can
+// ensure topic claim semantics are adhered to. In topic claim mode this will be called
+// by every claim during a release.
+func (c *Consumer) claimTerminated(cl *claim, released bool) {
+	// For now, we don't care except in the topic claim mode
+	if !c.options.ClaimEntireTopic {
+		return
+	}
+
+	latestClaims, claimsErr := c.GetCurrentTopicClaims()
+	if claimsErr != nil {
+		// This only impacts update topic claims later
+		log.Errorf("claimTerminated: failed getting current claims: %s", claimsErr)
+	}
+	delete(latestClaims, cl.topic)
```

Comment:
I think the way updateTopicClaims is written today, it assumes things are being updated by a single thread since it receives the full copy. This is true because we have a single thread calling ClaimTopics today (and fast claim initially). With this implementation, since claimTerminated is called by different go routines, we can race here where calling c.GetCurrentTopicClaims followed by updateTopicClaims can race! I believe you have to take a full Lock for the duration of both operations to make sure we serialize things (in case different topics get released simultaneously).